File: src/lib/user-memory-maintenance.ts
import { getTransport } from "./transport"

export interface MemoryReview {
  id: string
  target: { id: string; contentDigest: string }
  content: string
  reason: string
  quote: string
  status: "pending" | "applied" | "dismissed"
  createdAt: string
  resolvedAt: string | null
}

export interface MemoryMaintenanceStatus {
  revision: string
  busy: boolean
  lastCheckedAt: string | null
  activeCount: number
  expiredCount: number
  staleReviewIds: string[]
  modelReview: {
    lastAttemptAt: string | null
    lastCompletedAt: string | null
    lastErrorCode: string | null
    reviews: MemoryReview[]
  }
}

export interface MemoryMigrationPreview {
  sourceRevision: string
  generatedAt: string
  counts: Record<string, number>
  warnings: string[]
  unparsedLines: MemoryMigrationIssue[]
  readyForShadowImport: boolean
  records: {
    id: string
    kind: string
    content: string | null
    state: string
    scopeType: string
    scopeKey: string
    validTo: string | null
  }[]
}

export interface MemoryMigrationIssue {
  lineNumber: number
  content: string | null
  contentDigest: string
  sensitive: boolean
}

export interface ReconcileMemoryMigrationResult {
  converted: number
  backupFile: string | null
}

export const getMemoryMaintenance = () =>
  getTransport().call<MemoryMaintenanceStatus>("get_user_memory_maintenance")

export const runMemoryMaintenance = () =>
  getTransport().call<void>("run_user_memory_maintenance")

export const resolveMemoryReview = (request: {
  id: string
  expectedRevision: string
  apply: boolean
}) => getTransport().call<void>("resolve_user_memory_review", { request })

export const previewMemoryMigration = () =>
  getTransport().call<MemoryMigrationPreview>("preview_user_memory_migration")

export const reconcileMemoryMigration = (expectedRevision: string) =>
  getTransport().call<ReconcileMemoryMigrationResult>(
    "reconcile_user_memory_migration",
    { request: { expectedRevision } }
  )

File: src-tauri/src/commands/user_memory_maintenance.rs
use crate::app_error::AppCommandError;
#[cfg(feature = "tauri-runtime")]
use crate::user_memory::UserMemoryService;
use crate::user_memory::{
    MemoryMaintenanceStatus, MemoryMigrationPreview, ReconcileMemoryMigrationRequest,
    ReconcileMemoryMigrationResult, ResolveMemoryReviewRequest,
};
#[cfg(feature = "tauri-runtime")]
use std::sync::Arc;

#[cfg_attr(feature = "tauri-runtime", tauri::command)]
pub async fn get_user_memory_maintenance(
    #[cfg(feature = "tauri-runtime")] service: tauri::State<'_, Arc<UserMemoryService>>,
) -> Result<MemoryMaintenanceStatus, AppCommandError> {
    #[cfg(feature = "tauri-runtime")]
    {
        service.memory_maintenance_status().await
    }
    #[cfg(not(feature = "tauri-runtime"))]
    {
        Err(AppCommandError::configuration_invalid("tauri-only command"))
    }
}

#[cfg_attr(feature = "tauri-runtime", tauri::command)]
pub async fn reconcile_user_memory_migration(
    #[cfg(feature = "tauri-runtime")] service: tauri::State<'_, Arc<UserMemoryService>>,
    request: ReconcileMemoryMigrationRequest,
) -> Result<ReconcileMemoryMigrationResult, AppCommandError> {
    #[cfg(feature = "tauri-runtime")]
    {
        service.reconcile_memory_migration(request).await
    }
    #[cfg(not(feature = "tauri-runtime"))]
    {
        let _ = request;
        Err(AppCommandError::configuration_invalid("tauri-only command"))
    }
}

#[cfg_attr(feature = "tauri-runtime", tauri::command)]
pub async fn run_user_memory_maintenance(
    #[cfg(feature = "tauri-runtime")] service: tauri::State<'_, Arc<UserMemoryService>>,
) -> Result<(), AppCommandError> {
    #[cfg(feature = "tauri-runtime")]
    {
        service.run_memory_maintenance().await
    }
    #[cfg(not(feature = "tauri-runtime"))]
    {
        Err(AppCommandError::configuration_invalid("tauri-only command"))
    }
}

#[cfg_attr(feature = "tauri-runtime", tauri::command)]
pub async fn resolve_user_memory_review(
    #[cfg(feature = "tauri-runtime")] service: tauri::State<'_, Arc<UserMemoryService>>,
    request: ResolveMemoryReviewRequest,
) -> Result<(), AppCommandError> {
    #[cfg(feature = "tauri-runtime")]
    {
        service.resolve_memory_review(request).await
    }
    #[cfg(not(feature = "tauri-runtime"))]
    {
        let _ = request;
        Err(AppCommandError::configuration_invalid("tauri-only command"))
    }
}

#[cfg_attr(feature = "tauri-runtime", tauri::command)]
pub async fn preview_user_memory_migration(
    #[cfg(feature = "tauri-runtime")] service: tauri::State<'_, Arc<UserMemoryService>>,
) -> Result<MemoryMigrationPreview, AppCommandError> {
    #[cfg(feature = "tauri-runtime")]
    {
        service.preview_memory_migration().await
    }
    #[cfg(not(feature = "tauri-runtime"))]
    {
        Err(AppCommandError::configuration_invalid("tauri-only command"))
    }
}
